### ISSUE TITLE
Publish livestreams in RSS feed

### ISSUE THREAD

**Issue by sgiraudot** [Chocobozzz/PeerTube#3786]
Currently, the livestreams do not seem to appear in the RSS feed. It would nice if they would, as it would for example allow users to announce their future livestreams via RSS.

__Example:__ I plan to do a livestream on Saturday, I create the livestream the day before, viewers who subscribed to the RSS of the channel get the link for the livestream and are thus aware that there's going to be a stream soon (announcing the date/time when it will actually be streamed could remain the responsibility of the user).

I imagine that you could object that links for livestream would be irrelevant in RSS feed once the stream is over, so maybe it should only be done when the replay is switched on (as, as far as I know, the link remains the same for stream and replay).

**Comment by rigelk** [comment]
How do you plan to inform the user of a specific time at which the livestream is going to start? How do you deal with streams that already ended?

**Comment by sgiraudot** [comment]
> How do you plan to inform the user of a specific time at which the livestream is going to start?

I don't think that would be done automatically, but users could announce it themselves. What I did when I tried (and realized it did not appear in the RSS) was to simply put in in the title ("_rendez-vous samedi à 11h_"). Of course that's not ideal but the point is that the link would pop in the RSS feed as soon as the page is created.

> How do you deal with streams that already ended?

That's probably a good argument to only publish livestreams _that are programmed to have a replay_ on RSS (that way the link should remain valid not matter when).

**Comment by rigelk** [comment]
> > How do you plan to inform the user of a specific time at which the livestream is going to start?
> 
> I don't think that would be done automatically, but users could announce it themselves. What I did when I tried (and realized it did not appear in the RSS) was to simply put in in the title ("_rendez-vous samedi à 11h_"). Of course that's not ideal but the point is that the link would pop in the RSS feed as soon as the page is created.

Unfortunately, this is less than not ideal: we cannot automate it, since it relies on a specific language (here english) and not a semantic marker.

Invoking @agates here: since you have some experience with `<podcast:media>`, do you agree with my analysis that bare RSS2 doesn't provide any semantics for this? At least with Media RSS, they provide an example of using Dublin Core:

```xml
<dcterms:valid>
    start=2002-10-13T09:00+01:00;
    end=2002-10-17T17:00+01:00;
    scheme=W3C-DTF
</dcterms:valid>
```

Maybe setting a start date as soon as it has started, and an end date when the livestream ends would be enough. Non-saved livestreams would be deleted from the feed directly after the stream ends.

**Comment by agates** [comment]
I'm not aware of something in place for this -- @daveajones, are you aware of any kind of prior art in this regard?  I know we've had a lot of discussion on how to represent live streams (https://github.com/Podcastindex-org/podcast-namespace/issues/181 for example), but nothing specific to an item.

**Comment by daveajones** [comment]
I'm not aware of anything with timing.  The rawvoice namespace had a live tag for this.  But no sense of metadata around timing I don't think.  I remember Yahoo having something too, but we determined it was weird I think?  I think you had discovered that one @agates .

But, this is a good use case for <podcast:media> to be aware of.  The "switching" is something we want to have a solution for, as one of the other use cases is switching between audio and video versions of a podcast seamlessly in the podcast app.  Switching between live and recording needs a solution if none exists (I really don't think it does).  And the timing to make it happen.

**Comment by agates** [comment]
@daveajones right, the yahoo thing is Media RSS as mentioned by @rigelk

I can see the idea of a "start time" for live content, which would allow sort of what you might see on youtube for an upcoming premier of a live stream video.  Having a predefined "end time" could make sense too for things that are estimated to only last so long.  That way the application wouldn't have to guess at the state, and for the purposes of UX can display it as "live" during that time period.

I think it could be done rather simply with something like ...
`<podcast:timeInterval>2021-03-04T11:00:00-06:00/PT3H<podcast:timeInterval>`
... for an item that starts at 11am CST on March 4, 2021 and last 3 hours -- using an [ISO 8601 time interval](https://en.wikipedia.org/wiki/ISO_8601#Time_intervals)

The issue is, isn't it best practice to avoid publishing the item more than once?  This may be best done outside of the of the item because it might not even have a URL for enclosure etc when first published.  May be another reason to argue for `podcast:media` on the channel level, or in the proposed `podcast:live` tag.

**Comment by agates** [comment]
To address this once again, two years later:

1. We have [`<podcast:liveItem>`](https://github.com/Podcastindex-org/podcast-namespace/blob/main/docs/1.0.md#live-item) which has ISO8601 start and end timestamps.  These timestamps are purely for UX purposes, for obvious reasons.
2. We have a decentralized notification system called [Podping](https://podping.org/) which podcast applications are already using for live stream notifications.  Currently the main implementation is with the Hive blockchain but we already have plans to add more backends.
   * PR #5487 adds a hook that can be used by plugins to react to live stream status changes, and I will be using this to write a Podping notification plugin.
